# 3_EURES_CH_Daten_reinigen

<font color='red'>In diesem Notebook werden die gescrapten Daten (abgespeichert als Data_EURES_CH_17022019.csv) mit Hilfe von Pandas gereinigt und klassifiziert.</font>

### Schritt 1: Daten laden

In [27]:
import pandas as pd

In [28]:
path = "Data_EURES_CH_17022019.csv"

with open(path) as f:
    dateiinhalt = f.read()

In [29]:
df = pd.read_csv(path)

### Schritt 2: Datenstruktur überblicken

In [39]:
df.shape

(4831, 11)

**ERKENNTNIS:** Die Tabelle hat 4831 Einträge, was der Anzahl Jobinserate aus der Schweiz entspricht

In [40]:
df.head(3)

id   creationDate  lastModificationDate  \
0  MDgyQ05KQyA5  1547103651000         1550419220000   
1  MDgyWFJLUSA5  1548745236000         1550419219000   
2  MDgyV0xYUyA5  1548662527000         1550417428000   

                                              title  \
0  Conseiller / Conseillère clientèle en assurances   
1                   CONSEILLER A LA CLIENTELE (H/F)   
2                  Téléconseiller / Téléconseillère   

                                         description  numberOfPosts  \
0  Prospection téléphonique <br>Suivi de la clien...              1   
1  Analyser la situation globale de vos clients a...              1   
2  Notre club s'oriente aujourd'hui vers un club ...              1   

      locationMap  euresFlag                         jobCategoriesCodes  \
0  {'CH': [None]}      False  ['http://data.europa.eu/esco/isco/C3321']   
1  {'CH': [None]}      False  ['http://data.europa.eu/esco/isco/C3321']   
2  {'CH': [None]}      False  ['http://data.europa.eu/esco/isco/C5244']   

                                            employer  score  
0  {'name': 'PHOENIX MANAGEMENT', 'website': None...    NaN  
1  {'name': 'MF CONSEIL SARL', 'website': None, '...    NaN  
2  {'name': 'FRANCE LOISIRS SUISSE', 'website': N...    NaN

In [41]:
df.tail(3)

id   creationDate  lastModificationDate  \
4828  MTAwMDAtMTE2NDk0NTI5OC1TIDE  1535144224976         1535144224976   
4829  MTAwMDAtMTE0ODA0MDA2NS1TIDE  1531547424942         1531547424942   
4830  MTAwMDAtMTE0Nzg5ODUwOC1TIDE  1531547424251         1531547424251   

                                                  title  \
4828  Gesundheits- und Krankenpfleger/in auf der Gyn...   
4829  Facharzt (m/w) Innere Medizin (Facharzt/-ärzti...   
4830  Chefarzt (m/w) Ophtalmologie (Facharzt/-ärztin...   

                                            description  numberOfPosts  \
4828  Stellenangebotsbeschreibung: <br>Die Personal ...              2   
4829  Stellenangebotsbeschreibung: <br>Die Klinik su...              1   
4830  Stellenangebotsbeschreibung: <br>Die Klinik su...              1   

         locationMap  euresFlag  \
4828  {'CH': [None]}      False   
4829  {'CH': [None]}      False   
4830  {'CH': [None]}      False   

                                     jobCategoriesCodes  \
4828          ['http://data.europa.eu/esco/isco/C3221']   
4829          ['http://data.europa.eu/esco/isco/C2212']   
4830  ['http://data.europa.eu/esco/isco/C2212', 'htt...   

                                               employer  score  
4828  {'name': 'Personal Spezial Arzt und Pflege Ver...    NaN  
4829  {'name': 'Connecting People Medical', 'website...    NaN  
4830  {'name': 'Connecting People Medical', 'website...    NaN

**ERKENNTNIS:** Die ersten und die letzten Einträge stimmen überein mit den Jobinseraten auf der ersten und letzten Page auf EURES. Der Scraper scheint alle Jobinserate zu scrapen.

In [42]:
df.dtypes

id                       object
creationDate              int64
lastModificationDate      int64
title                    object
description              object
numberOfPosts             int64
locationMap              object
euresFlag                  bool
jobCategoriesCodes       object
employer                 object
score                   float64
dtype: object

Entscheidend für die Frage, welche Schweizer Jobs im Ausland ausgeschrieben werden sind die Variablen **jobCategoriesCodes**  und **numberOfPosts**.

In [102]:
df['jobCategoriesCodes'].head(100).tail(20)

80    ['http://data.europa.eu/esco/isco/C7112', 'htt...
81    ['http://data.europa.eu/esco/isco/C5131', 'htt...
82            ['http://data.europa.eu/esco/isco/C5151']
83            ['http://data.europa.eu/esco/isco/C3118']
84    ['http://data.europa.eu/esco/isco/C7233', 'htt...
85            ['http://data.europa.eu/esco/isco/C7126']
86    ['http://data.europa.eu/esco/isco/C3118', 'htt...
87    ['http://data.europa.eu/esco/isco/C7421', 'htt...
88            ['http://data.europa.eu/esco/isco/C2654']
89            ['http://data.europa.eu/esco/isco/C2212']
90            ['http://data.europa.eu/esco/isco/C3115']
91            ['http://data.europa.eu/esco/isco/C7123']
92            ['http://data.europa.eu/esco/isco/C7123']
93    ['http://data.europa.eu/esco/isco/C7412', 'htt...
94    ['http://data.europa.eu/esco/isco/C7112', 'htt...
95    ['http://data.europa.eu/esco/isco/C2212', 'htt...
96            ['http://data.europa.eu/esco/isco/C4120']
97            ['http://data.europa.eu/esco/isco/

**ERKENNTNIS:** Wohin führen die Links unter jobCategoriesCodes? Hinter jedem Link (http://data.europa.eu/esco/isco/C7212?) ist auf der Webseite von Eures ein Jobprofil hinterlegt nach dem International Standard Classification of Occupations - ISCO 08 (https://www.bfs.admin.ch/bfs/de/home/statistiken/arbeit-erwerb/nomenclaturen/isco08.html). Damit lassen sich die Daten klassifzieren.

In [43]:
df.describe()

creationDate  lastModificationDate  numberOfPosts  score
count  4.831000e+03          4.831000e+03    4831.000000    0.0
mean   1.546273e+12          1.548625e+12       1.587663    NaN
std    5.156601e+09          1.661317e+09       1.819530    NaN
min    1.531547e+12          1.531547e+12       1.000000    NaN
25%    1.545318e+12          1.548108e+12       1.000000    NaN
50%    1.548223e+12          1.549062e+12       1.000000    NaN
75%    1.549323e+12          1.549922e+12       1.000000    NaN
max    1.550357e+12          1.550419e+12      40.000000    NaN

In [44]:
df['numberOfPosts'].value_counts().sort_index()

1     4052
2      132
3      170
4      150
5      208
6       14
7        8
8        9
9        3
10      70
11       2
12       4
15       5
20       1
30       2
40       1
Name: numberOfPosts, dtype: int64

In [45]:
df[df['numberOfPosts'] == 40]

id   creationDate  lastModificationDate  \
2361  MTAwMDAtMTE2OTExMTU5NS1TIDE  1549063740383         1549063740383   

                                                  title  \
2361  Schweisser/Rohrschlosser/Anlagenbau (Schweißer...   

                                            description  numberOfPosts  \
2361  Stellenangebotsbeschreibung: <br>- Isometrie-R...             40   

         locationMap  euresFlag                         jobCategoriesCodes  \
2361  {'CH': [None]}      False  ['http://data.europa.eu/esco/isco/C7212']   

                                               employer  score  
2361  {'name': 'inova Personal Amriswil GmbH', 'webs...    NaN

In [46]:
df['jobCategoriesCodes'].tail(2470).head(1)

2361    ['http://data.europa.eu/esco/isco/C7212']
Name: jobCategoriesCodes, dtype: object

**ERKENNTNIS:** Die Variable numberOfPosts ist ein Zahlenwert (integer). Sie hat einen Maximalwert von 40. Firmen scheinen zum Teil mehr als eine Stelle unter einem Jobinserat zu inserieren. Eine Firma hat sogar 40 Stellen in einem Jobinserat aufgeschalten. Insgesamt sind unter den 4831 Jobinseraten 7670 offene Stellen subsumiert. Auch wenn 40 Stellen ausgeschrieben sind, können diese alle das gleiche Jobprofil haben.

**AUSNAHME 1:** Weiter Aufschluss geben können die Direktlinks zu den Jobinseraten. Dazu muss nur die id, indiesem Fall MTAwMDAtMTE2OTExMTU5NS1TIDE, in der Url eines beliebigen Jobinserates ausgewechselt werden:  https://ec.europa.eu/eures/eures-searchengine/page/main?lang=en&app=2.1.1p1-build-0#/jv-details/MTAwMDAtMTE2OTExMTU5NS1TIDE

Es handelt sich offenbar um einen Jobvermittler welcher 40 Schlosser sucht.

**AUSNAHME 2:** Gucken wir uns einen weiteren Ausnahmefall an: In einem Jobinserat sind mehr als ein Jobprofil hinterlegt: https://ec.europa.eu/eures/eures-searchengine/page/main?lang=en&app=2.1.1p1-build-0#/jv-details/MTAwMDAtMTE2ODc2OTY1NS1TIDE

Es werden 3 Intesivpfleger gesucht, der Jobinserent hat aber 2 Jobprofile hinterlegt: Health care assistants und Nursing associate professionals.

<font color='red'>**WEITERES VORGEHEN**: Es spielt keine Rolle ob in der Spalte 'jobCategoriesCodes' nur ein oder mehrere Links zu Jobkategorien hinterlegt sind. Wird nur die erste Hierarchiestufe gebraucht, lassen sich die Inserate einfach kategorisieren. Es braucht eine neue Spalte'jobCodes10'wo die einzelnen Inserate der obersten Hierarchiestufe gemäss ISCO 08-Standard zugewiesen werden.</font>

### Schritt 3: Daten reinigen

**VERSUCH 1:** Die Variable 'jobCategoriesCodes' nach dem Komma splitten. (siehe dazu https://stackoverflow.com/questions/14745022/how-to-split-a-column-into-two-columnsDie))

In [6]:
df[['jobCategoriesCodes1','jobCategoriesCodes2']] = pd.DataFrame(df.jobCategoriesCodes.values.tolist(), index= df.index)
print (df)

ValueError: Columns must be same length as key

**Funktioniert nicht!** Warum? Unter 'jobCategoriesCodes' sind Listen mit eintweder einem oder mehreren Elementen abgespeichert. Damit funktioniert obiger Befehl nicht.

**VERSUCH 2:** Die Variable 'jobCategoriesCodes' umwandeln in einen String und dann splitten.

In [7]:
import json

In [8]:
df['jobCategoriesCodes'][93]

"['http://data.europa.eu/esco/isco/C7412', 'http://data.europa.eu/esco/isco/C7411']"

In [11]:
jobCode = json.loads(df['jobCategoriesCodes'][93].replace("'",'"'))[0]
jobCode

'http://data.europa.eu/esco/isco/C7412'

Es funktioniert. Die Variable 'jobCategoriesCodes' ist immer eine Url und gleich aufgebaut:  

http://data.europa.eu/esco/isco/C.....

Man kann also nach '/' splitten, das letzte Element auswählen C7412 und dann die erste Zahl nach dem C.

In [12]:
jobCodeURL = json.loads(df['jobCategoriesCodes'][93].replace("'",'"'))[0]
jobCodeURL

'http://data.europa.eu/esco/isco/C7412'

In [18]:
jobCodeNumber = json.loads(df['jobCategoriesCodes'][93].replace("'",'"'))[0].split("/")[-1][1]
jobCodeNumber

'7'

Jetzt die Funktion auf die ganze Spalte jobCategoriesCodes anwenden und in einer neuen Spalte 'jobCodes10' abspeichern.

In [19]:
df['jobCodes10'] = json.loads(df['jobCategoriesCodes'].replace("'",'"'))[0].split("/")[-1][1]

TypeError: the JSON object must be str, bytes or bytearray, not Series

**Funktioniert nicht!** Warum? json.loads lässt sich nicht auf eine Series anwenden. Es braucht einen Loop.

In [20]:
list_new = []

for var in df['jobCategoriesCodes']:
    list_new.append(json.loads(var.replace("'",'"'))[0].split("/")[-1][1])
list_new


['3',
 '3',
 '5',
 '9',
 '7',
 '7',
 '7',
 '7',
 '8',
 '5',
 '7',
 '9',
 '7',
 '7',
 '8',
 '3',
 '7',
 '7',
 '5',
 '9',
 '5',
 '7',
 '5',
 '7',
 '1',
 '7',
 '7',
 '2',
 '1',
 '2',
 '8',
 '7',
 '7',
 '8',
 '2',
 '7',
 '7',
 '8',
 '4',
 '7',
 '5',
 '2',
 '2',
 '2',
 '2',
 '2',
 '2',
 '2',
 '2',
 '2',
 '2',
 '1',
 '7',
 '7',
 '7',
 '7',
 '1',
 '5',
 '7',
 '7',
 '1',
 '1',
 '1',
 '4',
 '7',
 '7',
 '2',
 '9',
 '8',
 '7',
 '7',
 '7',
 '7',
 '7',
 '7',
 '3',
 '3',
 '7',
 '7',
 '4',
 '7',
 '5',
 '5',
 '3',
 '7',
 '7',
 '3',
 '7',
 '2',
 '2',
 '3',
 '7',
 '7',
 '7',
 '7',
 '2',
 '4',
 '5',
 '3',
 '7',
 '7',
 '7',
 '2',
 '2',
 '2',
 '7',
 '1',
 '3',
 '7',
 '7',
 '8',
 '2',
 '2',
 '7',
 '7',
 '2',
 '2',
 '4',
 '7',
 '9',
 '8',
 '7',
 '7',
 '2',
 '1',
 '1',
 '4',
 '6',
 '5',
 '7',
 '2',
 '3',
 '5',
 '7',
 '2',
 '5',
 '7',
 '9',
 '4',
 '7',
 '1',
 '4',
 '2',
 '2',
 '2',
 '1',
 '2',
 '3',
 '7',
 '5',
 '7',
 '5',
 '5',
 '2',
 '2',
 '3',
 '2',
 '2',
 '5',
 '5',
 '3',
 '7',
 '1',
 '1',
 '2',
 '5',
 '1'

In [22]:
df['jobCodes10'] = list_new

In [26]:
df

id   creationDate  \
0                                          MDgyQ05KQyA5  1547103651000   
1                                          MDgyWFJLUSA5  1548745236000   
2                                          MDgyV0xYUyA5  1548662527000   
3                                          MDgzUVZCUSA5  1550053870000   
4                                          MDgxR0pZQiA5  1544693437000   
5                                          MDgzSk5RTSA5  1549467071000   
6                                          MDgyUUJTRCA5  1548082867000   
7                                          MDgzSk1MRyA5  1549465268000   
8                                          MDgzQ0xSUSA5  1549008106000   
9                                          MDgzUk5YUyA5  1550129481000   
10                                         MDgyWkZCWSA5  1548837159000   
11                                         MDgyWlBNWSA5  1548928881000   
12                                         MDgyTVpYWiA5  1547807702000   
13                                         MDgzUUpSUiA5  1550044926000   
14                                         MDgzRFRYSyA5  1549261842000   
15                                         MDgzRkhQSCA5  1549274502000   
16                                         MDgzUUpGUiA5  1550043056000   
17                                         MDgzUUpHSyA5  1550043046000   
18                                         MDgzR0ZRSiA5  1549350020000   
19                                         MDgzR1hQRyA5  1549371651000   
20                                         MDgzR1lNSyA5  1549443687000   
21                                         MDgyU01LTiA5  1548252106000   
22                                         MDgyUktOSCA5  1548169260000   
23                          MTAwMDAtMTE2OTUyMTQ5NS1TIDE  1550356787864   
24                               MTI4MTEtMzAzMjk5LVMgMQ  1550356774432   
25                          MTAwMDAtMTE2OTUxODY4Ny1TIDE  1550356760065   
26                          MTAwMDAtMTE2OTUxODU3NC1TIDE  1550356759584   
27                          MTAwMDAtMTE2OTUxODA3OS1TIDE  1550356753314   
28                           MTI5NjMtMjUzODExNDc2LVMgMQ  1550356752115   
29                               MTI4MTEtMzAzMTc5LVMgMQ  1550356730640   
...                                                 ...            ...   
4801  YmNkNmEyNWItZTgxNS0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001975   
4802  M2UyMTgzMDgtZThkZC0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001910   
4803  OWIxZmNkMDktZThiNC0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001910   
4804  Yzk4M2YwNTEtZThiMi0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001910   
4805  ZDBkNjg1NjUtZTlhNy0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001677   
4806  YTJjYzY5Y2YtZThiNS0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001598   
4807  MzA3ZWQ4NzUtYWY2Zi0xMWU4LWEzOGMtMDA1MDU2YWMwOD...  1544175001596   
4808  MDJkNDZjMDgtZThhYS0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001577   
4809  ZTNlOTgzYmUtZWJjYi0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001544   
4810  NGE0ZjAyYjYtZTc3Mi0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001537   
4811  YzU0N2U4YWEtZTliMC0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001530   
4812  M2U3ODQzODUtZWIyYi0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001530   
4813  NjU2Y2FlYTEtZTlhZi0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001526   
4814  ZDJlOTE4YzQtZWJjYi0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001435   
4815  YzJjMWM1YmItZWJjYS0xMWU4LWIwMDktMDA1MDU2YWMwOD...  1544175001427   
4816  MGJhOTNlNTEtZTQxZi0xMWU4LWIxNzUtMDA1MDU2YWMwOD...  1544175001301   
4817  MWEyZjM5OWQtZjk1Yy0xMWU4LTk3N2MtMDA1MDU2YWMwOD...  1544175000990   
4818                        MTAwMDAtMTE2NzI2ODc2NS1TIDE  1542840878637   
4819                        MTAwMDAtMTE2NzI1MzM2NS1TIDE  1542756384569   
4820                        MTAwMDAtMTE2NzI2ODkzNS1TIDE  1542840878909   
4821                        MTAwMDAtMTE2NzI0ODQxMC1TIDE  1542756343650   
4822                        MTAwMDAtMTE2NzI0NjAzMC1TIDE  1542756329001   
4823                           MDM5NzE4MTEwMDAwMTYgMjYx  

In [23]:
df.to_csv('Data_jobCodes10.csv', index=False)

<font color='red'>Weiter geht es mit dem **Jupyter Notebook: 4_EURES_CH_Pandas_Plotten**.</font>